tutorial/lesson_19_staging_func_or_image_param.cpp

```diff
@@ -0,0 +1,289 @@
+// Halide tutorial lesson 19: Staging Func or ImageParam
+
+// This lesson demonstrates how to use Func::in() and ImageParam::in() to
+// schedule a Func differently in different places and to stage loads from
+// a Func or an ImageParam.
+
+// On linux, you can compile and run it like so:
+// g++ lesson_19*.cpp -g -I ../include -L ../bin -lHalide -lpthread -ldl -o lesson_19 -std=c++11
+// LD_LIBRARY_PATH=../bin ./lesson_19
+
+// On os x:
+// g++ lesson_19*.cpp -g -I ../include -L ../bin -lHalide -o lesson_19 -std=c++11
+// DYLD_LIBRARY_PATH=../bin ./lesson_19
+
+// If you have the entire Halide source tree, you can also build it by
+// running:
+//    make tutorial_lesson_19_staging_func_or_image_param
+// in a shell with the current directory at the top of the halide
+// source tree.
+
+// The only Halide header file you need is Halide.h. It includes all of Halide.
+#include "Halide.h"
+
+// We'll also include stdio for printf.
+#include <stdio.h>
+
+using namespace Halide;
+
+int main(int argc, char **argv) {
+    // First we'll declare some Vars to use below.
+    Var x("x"), y("y");
+
+    // This lesson will be about "wrapping" a Func or an ImageParam using the
+    // Func::in() and ImageParam::in() directives, which may be useful for
+    // variety of use cases.
+
```

Comment:
I think this should start by demonstrating what in() does to a Halide pipeline by showing equivalent Halide code with and without in, and then explain that this can be used for a variety of scheduling tricks. Each section below is then one of the scheduling tricks that you can use in() for.